### ISSUE TITLE
MLproject1: Diabetes Detection Model

### ISSUE THREAD

**Issue by nnfuad** [nnfuad/fuados#1]
# Diabetes Prediction using PyTorch

## Project Overview
This project aims to build a **machine learning model** to predict whether a patient has diabetes based on various health measurements. The dataset is sourced from  **Kaggle**.

The project demonstrates:
- Data loading and preprocessing
- Feature scaling and train-test splitting
- Building a simple **feedforward neural network** in PyTorch
- Model training and evaluation
- Visualization of training loss

## Dataset
- **Source:** An open-source CSV type diabetes [datasheet](https://www.kaggle.com/datasets/saurabh00007/diabetescsv) from Kaggle.
- **Features:** Pregnancies, Glucose, BloodPressure, SkinThickness, Insulin, BMI, DiabetesPedigreeFunction, Age
- **Target:** Outcome (1 if diabetic, 0 otherwise)

## Tools and Libraries
- Python
- PyTorch
- Pandas
- NumPy
- Matplotlib
- Scikit-learn

### 1. Load Dataset
```python
import pandas as pd

df = pd.read_csv('diabetes.csv')
print(df.head())
```

### 2. Explore Data
```python
print(df.info())
print(df.describe())
```

### 3. Preprocess Data
- Handle missing values (if any)
- Split features and labels
- Standardize features

```python
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

X = df.drop('Outcome', axis=1).values
y = df['Outcome'].values

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)
```

### 4. Convert Data to PyTorch Tensors
```python
import torch

X_train_tensor = torch.tensor(X_train, dtype=torch.float32)
y_train_tensor = torch.tensor(y_train, dtype=torch.float32).view(-1,1)
X_test_tensor = torch.tensor(X_test, dtype=torch.float32)
y_test_tensor = torch.tensor(y_test, dtype=torch.float32).view(-1,1)
```

### 5. Build the Model
```python
import torch.nn as nn

class SimpleNN(nn.Module):
    def __init__(self):
        super().__init__()
        self.layers = nn.Sequential(
            nn.Linear(8, 16),
            nn.ReLU(),
            nn.Linear(16, 8),
            nn.ReLU(),
            nn.Linear(8, 1),
            nn.Sigmoid()
        )
    def forward(self, x):
        return self.layers(x)

model = SimpleNN()
```

### 6. Define Loss and Optimizer
```python
criterion = nn.BCELoss()
optimizer = torch.optim.Adam(model.parameters(), lr=0.001)
```

### 7. Train the Model
```python
import matplotlib.pyplot as plt

loss_values = []
epochs = 100

for epoch in range(epochs):
    optimizer.zero_grad()
    outputs = model(X_train_tensor)
    loss = criterion(outputs, y_train_tensor)
    loss.backward()
    optimizer.step()
    
    loss_values.append(loss.item())
    if (epoch+1) % 10 == 0:
        print(f'Epoch {epoch+1}, Loss: {loss.item():.4f}')

plt.plot(range(epochs), loss_values)
plt.xlabel('Epoch')
plt.ylabel('Loss')
plt.title('Training Loss over Epochs')
plt.show()
```

### 8. Evaluate Model
```python
from sklearn.metrics import accuracy_score

with torch.no_grad():
    y_pred = model(X_test_tensor)
    y_pred_class = (y_pred > 0.5).float()
    acc = accuracy_score(y_test_tensor, y_pred_class)
    print(f'Test Accuracy: {acc*100:.2f}%')
```

## Results
- Training loss decreases over epochs
- Test accuracy typically around **75-85%** depending on hyperparameters
- Confusion matrix or ROC curve can be added for detailed evaluation

## Future Improvements
- Tune hyperparameters (learning rate, layers, neurons)
- Add cross-validation
- Explore feature selection
- Implement more advanced models like Random Forest or XGBoost

## How to Run
1. Clone the repository
2. Install required libraries
3. Run the Jupyter notebook or Python script

## Tech Stack
- Python
- PyTorch
- Pandas
- NumPy
- Matplotlib
- Scikit-learn

---
**Author:** Nur Nafis Fuad
**Date:** 2004-11-08
**GitHub:** [nnfuad](https://github.com/nnfuad/)
